File: library/src/main/java/com/nshmura/strictmodenotifier/StrictModeReportDetailActivity.java
package com.nshmura.strictmodenotifier;

import android.annotation.TargetApi;
import android.app.Activity;
import android.content.ClipData;
import android.content.ClipboardManager;
import android.content.Context;
import android.content.Intent;
import android.graphics.Color;
import android.os.Build;
import android.os.Bundle;
import android.text.Spannable;
import android.text.SpannableString;
import android.text.style.ForegroundColorSpan;
import android.text.style.StyleSpan;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.widget.TextView;
import android.widget.Toast;

public class StrictModeReportDetailActivity extends Activity {

  private static final String EXTRA_REPORT = "EXTRA_REPORT";

  private ViolationStore violationStore;
  private StrictModeViolation report;

  @Override protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    setContentView(R.layout.strictmode_notifier_activity_report_detail);

    violationStore = new ViolationStore(StrictModeReportDetailActivity.this);

    report = (StrictModeViolation) getIntent().getSerializableExtra(EXTRA_REPORT);

    TextView stackTreace = (TextView) findViewById(R.id.__stacktrace_text);

    final String stacktraceText = report.getStacktraceText();
    final Spannable span = new SpannableString(stacktraceText);

    setColor(span, stacktraceText, getPackageName(), Color.WHITE);

    stackTreace.setText(span);

    if (report.violationType != null) {
      ReportActivityUtils.setTitle(this,
          ViolationTypeInfo.convert(report.violationType).violationName());
      ReportActivityUtils.setDisplayHomeAsUpEnabled(this, true);
    }

    findViewById(R.id.__delete_button).setOnClickListener(new View.OnClickListener() {
      @Override public void onClick(View v) {
        violationStore.remove(report);
        finish();
      }
    });
  }

  @Override public boolean onMenuItemSelected(int featureId, MenuItem item) {
    switch (item.getItemId()) {
      case android.R.id.home:
        finish();
        break;
    }
    return super.onMenuItemSelected(featureId, item);
  }

  @Override public boolean onCreateOptionsMenu(Menu menu) {
    MenuInflater inflater = getMenuInflater();
    inflater.inflate(R.menu.strictmode_notifier_menu, menu);

    if (Build.VERSION.SDK_INT < Build.VERSION_CODES.HONEYCOMB) {
      menu.removeItem(R.id.__menu_copy);
    }

    return true;
  }

  @Override public boolean onOptionsItemSelected(MenuItem item) {
    int itemId = item.getItemId();
    if (itemId == R.id.__menu_copy) {
      copy();
      return true;
    } else if (itemId == R.id.__menu_share) {
      share();
      return true;
    } else {
      return super.onOptionsItemSelected(item);
    }
  }

  @TargetApi(Build.VERSION_CODES.HONEYCOMB) private void copy() {
    ClipData clip = ClipData.newPlainText("info", shareText());
    ClipboardManager clipboard = (ClipboardManager) getSystemService(Context.CLIPBOARD_SERVICE);
    clipboard.setPrimaryClip(clip);
    Toast.makeText(this, R.string.strictmode_notifier_copyped, Toast.LENGTH_LONG).show();
  }

  private void share() {
    Intent sendIntent = new Intent();
    sendIntent.setAction(Intent.ACTION_SEND);
    sendIntent.putExtra(Intent.EXTRA_TEXT, shareText());
    sendIntent.setType("text/plain");

    try {
      startActivity(Intent.createChooser(sendIntent,
          getResources().getText(R.string.strictmode_notifier_menu_share)));
    } catch (Exception e) {
      e.printStackTrace();
    }
  }

  private String shareText() {
    String shareText;
    if (report.violationType != null) {
      shareText = ViolationTypeInfo.convert(report.violationType).violationName();
      shareText += "\n\n" + report.getStacktraceText();
    } else {
      shareText = report.getStacktraceText();
    }
    return shareText;
  }

  public static void start(Context context, StrictModeViolation report) {
    context.startActivity(createIntent(context, report));
  }

  public static Intent createIntent(Context context, StrictModeViolation report) {
    Intent intent = new Intent(context, StrictModeReportDetailActivity.class);
    intent.putExtra(EXTRA_REPORT, report);
    return intent;
  }

  public void setColor(Spannable span, String stacktraceText, String hilight, int color) {
    final int length = hilight.length();

    int start = 0;
    while (true) {
      int index = stacktraceText.indexOf(hilight, start);
      if (index < 0) {
        break;
      }

      span.setSpan(new StyleSpan(android.graphics.Typeface.BOLD), index, index + length,
          Spannable.SPAN_INCLUSIVE_INCLUSIVE);

      span.setSpan(new ForegroundColorSpan(color), index, index + length,
          Spannable.SPAN_INCLUSIVE_INCLUSIVE);

      start = index + length;
    }
  }
}

File: notifier/src/main/java/io/github/juanimoli/strictmode/notifier/StrictModeReportDetailActivity.kt
package io.github.juanimoli.strictmode.notifier

import android.app.Activity
import android.content.ClipData
import android.content.ClipboardManager
import android.content.Context
import android.content.Intent
import android.graphics.Color
import android.graphics.Typeface
import android.os.Bundle
import android.text.Spannable
import android.text.SpannableString
import android.text.style.ForegroundColorSpan
import android.text.style.StyleSpan
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.TextView
import android.widget.Toast
import io.github.juanimoli.strictmode.notifier.commons.StrictModeViolation

class StrictModeReportDetailActivity : Activity() {
    private var violationStore: ViolationStore? = null
    private var report: StrictModeViolation? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.strictmode_notifier_activity_report_detail)

        violationStore = ViolationStore(this@StrictModeReportDetailActivity)
        report = intent.getSerializableExtra(EXTRA_REPORT) as StrictModeViolation?

        val stackTrace = findViewById<TextView>(R.id.__stacktrace_text)
        val stacktraceText = report?.stacktraceText
        val span: Spannable = SpannableString(stacktraceText)
        setColor(span, stacktraceText, packageName, Color.WHITE)
        stackTrace.text = span

        if (report?.violationType != null) {
            ReportActivityUtils.setTitle(
                this,
                ViolationTypeInfo.convert(report!!.violationType).violationName()
            )
            ReportActivityUtils.setDisplayHomeAsUpEnabled(this, true)
        }

        findViewById<View>(R.id.__delete_button).setOnClickListener {
            violationStore!!.remove(report)
            finish()
        }
    }

    override fun onMenuItemSelected(featureId: Int, item: MenuItem): Boolean {
        if (item.itemId == android.R.id.home) {
            finish()
        }

        return super.onMenuItemSelected(featureId, item)
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        val inflater = menuInflater
        inflater.inflate(R.menu.strictmode_notifier_menu, menu)

        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            R.id.__menu_copy -> {
                copy()
                true
            }
            R.id.__menu_share -> {
                share()
                true
            }
            else -> {
                super.onOptionsItemSelected(item)
            }
        }
    }

    private fun copy() {
        val clip = ClipData.newPlainText("info", shareText())
        val clipboard = getSystemService(CLIPBOARD_SERVICE) as ClipboardManager
        clipboard.setPrimaryClip(clip)

        Toast.makeText(this, R.string.strictmode_notifier_copyped, Toast.LENGTH_LONG).show()
    }

    private fun share() {
        val sendIntent = Intent()
        sendIntent.action = Intent.ACTION_SEND
        sendIntent.putExtra(Intent.EXTRA_TEXT, shareText())
        sendIntent.type = "text/plain"
        try {
            startActivity(
                Intent.createChooser(
                    sendIntent,
                    resources.getText(R.string.strictmode_notifier_menu_share)
                )
            )
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    private fun shareText(): String? {
        var shareText: String?
        if (report!!.violationType != null) {
            shareText = ViolationTypeInfo.convert(report!!.violationType).violationName()
            shareText += """
                
                
                ${report?.stacktraceText}
                """.trimIndent()
        } else {
            shareText = report?.stacktraceText
        }
        return shareText
    }

    @Suppress("MemberVisibilityCanBePrivate")
    fun setColor(span: Spannable, stacktraceText: String?, highlight: String, color: Int) {
        val length = highlight.length
        var start = 0

        while (true) {
            val index = stacktraceText!!.indexOf(highlight, start)
            if (index < 0) {
                break
            }

            span.setSpan(
                StyleSpan(Typeface.BOLD), index, index + length,
                Spannable.SPAN_INCLUSIVE_INCLUSIVE
            )

            span.setSpan(
                ForegroundColorSpan(color), index, index + length,
                Spannable.SPAN_INCLUSIVE_INCLUSIVE
            )

            start = index + length
        }
    }

    companion object {
        private const val EXTRA_REPORT = "EXTRA_REPORT"

        fun start(context: Context, report: StrictModeViolation?) {
            context.startActivity(createIntent(context, report))
        }

        @Suppress("MemberVisibilityCanBePrivate")
        fun createIntent(context: Context?, report: StrictModeViolation?): Intent {
            val intent = Intent(context, StrictModeReportDetailActivity::class.java)
            intent.putExtra(EXTRA_REPORT, report)
            return intent
        }
    }
}
